integration/version_test.go

```diff
@@ -0,0 +1,146 @@
+func TestVersionCheck(t *testing.T) {
+	lib.SetInsecureDevMode(true)
+	defer lib.SetInsecureDevMode(false)
+
+	var err error
+
+	authCfg := servicecfg.MakeDefaultConfig()
+	authCfg.Version = defaults.TeleportConfigVersionV3
+	authCfg.SetAuthServerAddress(utils.NetAddr{AddrNetwork: "tcp", Addr: "localhost:0"})
+	authCfg.Auth.Enabled = true
+	authCfg.Auth.ClusterName, err = services.NewClusterNameWithRandomID(types.ClusterNameSpecV2{
+		ClusterName: "auth-server",
+	})
+	require.NoError(t, err)
+	authCfg.Auth.ListenAddr.Addr = "localhost:0"
+	authCfg.Auth.NetworkingConfig.SetProxyListenerMode(types.ProxyListenerMode_Multiplex)
+
+	authCfg.Proxy.Enabled = false
+	authCfg.SSH.Enabled = false
+
+	launchFunc := func(ctx context.Context, authCfg servicecfg.Config, authRunErrCh chan error, signal chan struct{}) {
+		authRunErrCh <- service.Run(ctx, authCfg, func(cfg *servicecfg.Config) (service.Process, error) {
+			proc, err := service.NewTeleport(cfg)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			if _, err = proc.GetIdentity(types.RoleAdmin); err != nil {
+				proc.Close()
+				return nil, trace.Wrap(err)
+			}
+			signal <- struct{}{}
+
+			return proc, nil
+		})
+	}
+
+	teleportVersion, err := semver.NewVersion(teleport.Version)
+	require.NoError(t, err)
+
+	tests := map[string]struct {
+		initialVersion  string
+		expectedVersion string
+		expectError     bool
+	}{
+		"first-launch": {
+			initialVersion:  "",
+			expectedVersion: teleport.Version,
+			expectError:     false,
+		},
+		"old-version-upgrade": {
+			initialVersion:  fmt.Sprintf("%d.0.0", teleportVersion.Major-1),
+			expectedVersion: teleport.Version,
+			expectError:     false,
+		},
+		"major-upgrade-fail": {
+			initialVersion:  fmt.Sprintf("%d.0.0", teleportVersion.Major-2),
+			expectedVersion: fmt.Sprintf("%d.0.0", teleportVersion.Major-2),
+			expectError:     true,
+		},
+	}
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			// Create temporary directory for the auth service.
+			authDir := t.TempDir()
+			authCfg.DataDir = authDir
+			authCfg.Auth.StorageConfig.Params = backend.Params{defaults.BackendPath: filepath.Join(authDir, defaults.BackendDir)}
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			processStorage, err := storage.NewProcessStorage(ctx, filepath.Join(authDir, teleport.ComponentProcess))
+			require.NoError(t, err)
+			if test.initialVersion != "" {
+				err := processStorage.WriteTeleportVersion(ctx, test.initialVersion)
+				require.NoError(t, err)
+			}
+
+			authRunErrCh := make(chan error, 1)
+			signal := make(chan struct{})
+			go launchFunc(ctx, *authCfg, authRunErrCh, signal)
+
+			timeout := time.After(time.Second * 30)
+			select {
+			case err := <-authRunErrCh:
+				if test.expectError {
+					assert.Error(t, err)
+				} else {
+					require.NoError(t, err)
+				}
+			case <-timeout:
+				t.Fatal("timed out waiting for fetching identity")
+			case <-signal:
+			}
```

Comment:
We have a pretty major problem with flaky tests in Teleport, due to tests starting and waiting for various services which we largely have under control now but still, is there a way to make this test not depend on time?

For example, why do we need to start auth? Can we just instantiate auth of version X, call "initAuthService" on it, then start another auth of version X+1, init, make sure it works, then another of version X+3, init, make sure it fails, something like that? There would be no potential flakiness and the test would probably be simpler as well.

Reply:
I need to additionally ensure that state for `RoleAdmin` created and this logic might be changed in time in initialization flow, thats why initially I choose to use integration test for such case. I check this state to identify if we have new install with empty database rather than existing one.

For instance if I add tests which trigger `initAuthService` only, we will never cover the case with `majorVersionConstraint` (this case covers situation: let say we have version 15, we know that storing information about version added in v17, and we are trying to update to v18, so if we check that we have any version in storage means that we have older version that <v17)

Reply:
I rewrite the test to use `initAuthService` instead